An extension shouldn't be able to close another extension's popup with a call to `openPopup()`.

AriYoung00 opened the pull request:
#### 5d3fc7e38622142b0f113277c99edeffc40cd2b3
<pre>
An extension shouldn&apos;t be able to close another extension&apos;s popup with a call to `openPopup()`.
<a href="https://webkit.org/b/275914">https://webkit.org/b/275914</a>
<a href="https://rdar.apple.com/127836084">rdar://127836084</a>

<a href="https://github.com/w3c/webextensions/issues/160">https://github.com/w3c/webextensions/issues/160</a>

Reviewed by Timothy Hatcher.

Adds a static member to WebExtensionContext which signifies whether or
not any extension has an open popup, along with other changes to
WebExtensionContext to update this variable.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIActionCocoa.mm:
(WebKit::WebExtensionContext::actionOpenPopup): Modified to check the
state of popupIsOpen
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::readyToPresentPopup): Modified to set the
state of s_isPopupOpen when a popup is presented
(WebKit::WebExtensionAction::closePopup): Modified to set the state of
s_isPopupOpen when a popup is closed
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::popupOpening): Setter for s_isPopupOpen
(WebKit::WebExtensionContext::popupClosed):  Setter for s_isPopupOpen
(WebKit::WebExtensionContext::popupIsOpen const): Getter for
s_isPopupOpen

Canonical link: <a href="https://commits.webkit.org/280422@main">https://commits.webkit.org/280422@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/988c1d0c0b853e26f89d2080e4a02355584971c3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56604 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35931 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9077 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60218 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/7045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43554 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7236 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/45853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/7045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58634 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/33778 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/48850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/26714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30559 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6048 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/6453 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61901 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/516 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/6560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/53111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/518 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/48909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/53111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8409 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/31759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/32845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/33929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/32591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -963,6 +967,8 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
     RetainPtr<_WKWebExtensionStorageSQLiteStore> m_sessionStorageStore;
     RetainPtr<_WKWebExtensionStorageSQLiteStore> m_syncStorageStore;
+
+    inline static bool s_hasOpenPopup { false };
```
As we talked about, this shouldn't be a static. It would be best to put this on `WebExtensionController` so the state isn't global, while still preventing multiple extensions in the same controller (profile in Safari) from opening a popup while another is already showing.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -542,6 +542,10 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
 
+    void popupOpening() { s_hasOpenPopup = true; };
+    void popupClosed() { s_hasOpenPopup = false; };
+    bool popupIsOpen() const { return s_hasOpenPopup; };
```
These would move to `WebExtensionController`, and can be two methods. `isShowingActionPopup()` and `setShowingActionPopup()` would be good names.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIActionCocoa.mm:
```diff
@@ -205,7 +205,7 @@
     }
 
-    if (getOrCreateAction(tab.get())->presentsPopup())
+    if (getOrCreateAction(tab.get())->presentsPopup() && !popupIsOpen())
```
This would move up to the top of this method, near the other error, and based on https://github.com/w3c/webextensions/issues/160#issuecomment-2145197651 and the recent [meeting notes](https://github.com/w3c/webextensions/blob/9ea71fcc666472b2ecc1df877cfea65211db110c/_minutes/2024-06-06-wecg.md) we agreed to throw an error in this case.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -882,6 +882,10 @@ - (void)_otherPopoverWillShow:(NSNotification *)notification
+    RefPtr extensionController = extensionContext()->extensionController();
+    if (extensionController) {
+        extensionController->setShowingActionPopup(true);
+    }
```
```c++;
if (RefPtr extensionController = extensionContext()->extensionController())
    extensionController->setShowingActionPopup(true);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -942,6 +946,11 @@ - (void)_otherPopoverWillShow:(NSNotification *)notification
+    RefPtr extensionController = extensionContext()->extensionController();
+    if (extensionController) {
+        extensionController->setShowingActionPopup(false);
+    }
```
```c++;
if (RefPtr extensionController = extensionContext()->extensionController())
    extensionController->setShowingActionPopup(false);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIActionCocoa.mm:
```diff
@@ -166,6 +166,12 @@
     }
 
+    RefPtr controller = extensionController();
+    if (controller && controller->isShowingActionPopup()) {
```
`extensionController()` should always be non-null here, since the extension is loaded and there are earlier checks outside here that test this.

```
if (extensionController()->isShowingActionPopup()) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -882,6 +882,9 @@ - (void)_otherPopoverWillShow:(NSNotification *)notification
 
     m_popupPresented = true;
+    RefPtr extensionController = extensionContext()->extensionController();
+    if (extensionController)
```
```suggestion

    if (RefPtr extensionController = extensionContext()->extensionController())
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -942,6 +945,10 @@ - (void)_otherPopoverWillShow:(NSNotification *)notification
     m_presentsPopupWhenReady = false;
 
+    RefPtr extensionController = extensionContext()->extensionController();
+    if (extensionController)
```
```suggestion
    if (RefPtr extensionController = extensionContext()->extensionController())
```

AriYoung00 committed:
An extension shouldn't be able to close another extension's popup with a call to `openPopup()`.
https://webkit.org/b/275914
rdar://127836084

https://github.com/w3c/webextensions/issues/160

Reviewed by Timothy Hatcher.

Adds a static member to WebExtensionContext which signifies whether or
not any extension has an open popup, along with other changes to
WebExtensionContext to update this variable.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIActionCocoa.mm:
(WebKit::WebExtensionContext::actionOpenPopup): Modified to check the
state of popupIsOpen
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::readyToPresentPopup): Modified to set the
state of s_isPopupOpen when a popup is presented
(WebKit::WebExtensionAction::closePopup): Modified to set the state of
s_isPopupOpen when a popup is closed
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::popupOpening): Setter for s_isPopupOpen
(WebKit::WebExtensionContext::popupClosed):  Setter for s_isPopupOpen
(WebKit::WebExtensionContext::popupIsOpen const): Getter for
s_isPopupOpen

Canonical link: https://commits.webkit.org/280422@main
